src/Mod/Sketcher/Gui/EditModeConstraintCoinManager.h

```diff
@@ -0,0 +1,253 @@
+/***************************************************************************
+ *   Copyright (c) 2021 Abdullah Tahiri <abdullah.tahiri.yo@gmail.com>     *
+ *                                                                         *
+ *   This file is part of the FreeCAD CAx development system.              *
+ *                                                                         *
+ *   This library is free software; you can redistribute it and/or         *
+ *   modify it under the terms of the GNU Library General Public           *
+ *   License as published by the Free Software Foundation; either          *
+ *   version 2 of the License, or (at your option) any later version.      *
+ *                                                                         *
+ *   This library  is distributed in the hope that it will be useful,      *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU Library General Public License for more details.                  *
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#ifndef SKETCHERGUI_EditModeConstraintCoinManager_H
+#define SKETCHERGUI_EditModeConstraintCoinManager_H
+
+#include <vector>
+#include <functional>
+
+#include <Base/Parameter.h>
+#include <App/Application.h>
+
+#include <Mod/Sketcher/App/GeoList.h>
+
+#include "EditModeCoinManagerParameters.h"
+
+class SbVec3f;
+class SoRayPickAction;
+class SoPickedPoint;
+class SbVec3s;
+
+namespace Base {
+    template< typename T >
+    class Vector3;
+
+    class Vector2d;
+
+    class Placement;
+}
+
+namespace Part {
+    class Geometry;
+}
+
+namespace Sketcher {
+    class Constraint;
+    class PropertyConstraintList;
+};
+
+namespace SketcherGui {
+
+class ViewProviderSketch;
+
+using GeoList = Sketcher::GeoList;
+using GeoListFacade = Sketcher::GeoListFacade;
+
+/** @brief      Class for managing the Edit mode coin nodes of ViewProviderSketch relating to constraints.
+ *  @details    To be documented.
+ *
+ */
+class SketcherGuiExport EditModeConstraintCoinManager
+{
+private:
+    /// Coin Node indices for constraints
+    enum class ConstraintNodePosition {
+        MaterialIndex = 0,
+        DatumLabelIndex = 0,

```

Comment:
Suspect value, 0 used twice in an enum

Reply:
It is valid and intended.

Some constraints are built having children 0 to be the MaterialIndex and others the DatumLabelIndex. This removes quite some magic numbers and awful macros. However, when the Constraints handling in coin is refactored (this is Werner's comment on the forum), it may be possible to improve it further.